### PR TITLE
feat(assetissuance): get holders, issue tokens, redeem tokens

### PR DESCRIPTION
**Pull request by sansan**
### Description

adds example script for 
- getting asset holders
- issuing new tokens
- redeeming tokens

### JIRA Link

https://polymesh.atlassian.net/browse/DA-455
### Checklist

- [x] Updated the Readme.md (if required) ?

### CODE REVIEW

File: src/examples/assetIssuance.ts [hunk 1]
```diff
@@ -0,0 +1,123 @@
+import { BigNumber } from '@polymeshassociation/polymesh-sdk';
+import { Asset } from '@polymeshassociation/polymesh-sdk/types';
+
+import { getClient } from '~/common/client';
+import { parseArgs } from '~/common/utils';
+
+enum SCRIPT_ACTIONS {
```
Comment (polymath-eric) [inline src/examples/assetIssuance.ts:7]: I think this stuff is too distracting in an example file. At the very least this should go at the bottom so its not the first thing people see.

I think its more consistent for the example to do all three actions, without needing extra code to execute single ones. Alternatively this could be 3 different examples.

File: src/examples/assetIssuance.ts [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+
+    console.log(`Holders of ${asset.ticker}: \n`);
+
+    data.forEach(({ identity, balance }) => {
+      console.log(`- Identity: ${identity.did}, Balance: ${balance.toFormat()}`);
+    });
+  };
+
+  /**
+   * Issue tokens for Asset
```
Comment (polymath-eric) [inline src/examples/assetIssuance.ts:43]: Normally I like TSDoc comments, but in this case I think they distract from the example.

Its information about the script, and not the actual usage of the SDK, so I would remove these comments from here.

File: src/common/utils.ts [hunk 3]
```diff
@@ -26,3 +26,19 @@ export function toHumanObject(obj: unknown): unknown {
...
+
+/**
+ * @hidden
+ */
+export function parseArgs<T>(args: string[]): T {
+  // parse args
+  const parsedArgs = args.reduce((acc, arg) => {
+    const [key, value] = arg.split('=');
+
+    return Object.assign(acc, {
```
Comment (polymath-eric) [inline src/common/utils.ts:38]: Why make a new object here?

The sole reference to the acc is in the reduce function. Since its not shared until reduce returns its OK to mutate

File: src/common/utils.ts [hunk 4]
```diff
@@ -26,3 +26,19 @@ export function toHumanObject(obj: unknown): unknown {
...
   return obj;
 }
+
+/**
+ * @hidden
+ */
+export function parseArgs<T>(args: string[]): T {
+  // parse args
+  const parsedArgs = args.reduce((acc, arg) => {
+    const [key, value] = arg.split('=');
```
Comment (polymath-eric) [inline src/common/utils.ts:36]: This is never validated to be the case.

**Review by polymath-eric — COMMENTED** [review]
I think the example itself should only contain code about using the SDK, and not try to be a CLI.

Here over 50% of the lines in the example is doing argument parsing and command dispatch.

My recommendation would be to make 3 examples and remove all the running script code. @prashantasdeveloper and @VictorVicente might have a different opinion on this.

FWIW I like the interactivity, but I think it detracts from the copy paste use case these are for.

File: src/common/utils.ts (hunk 3 above)
Comment (polymath-eric) [reply src/common/utils.ts:38]: I misremembered how assign work here BTW

What you have makes sense since its just copying over to an existing object.

I worked on a code base where the pattern was `Object.assign({}, ...stuff)` which I was thinking of

**Comment by prashantasdeveloper** [comment]
> I think the example itself should only contain code about using the SDK, and not try to be a CLI.
> 
> Here over 50% of the lines in the example is doing argument parsing and command dispatch.
> 
> My recommendation would be to make 3 examples and remove all the running script code, or make it one big example. @prashantasdeveloper and @VictorVicente might have a different opinion on this.
> 
> FWIW I like the interactivity, but I think it detracts from the copy paste use case these are for.

I like the idea of splitting the code into separate examples. What we can do is create small snippets for now with almost no CLI interaction and later if we want we can create some scenario examples where we can call these methods with apt arguments. WDUT @polymath-eric ?

Just to be clear, I'm OK with the current implementation as well.

**Comment by polymath-eric** [comment]
I think the smaller snippets would keep it a bit more consistent, and should provide a similar experience to the script code.

I think the script code is at risk of being repeated or changing. 

I'd prefer only publishing patterns we are confident about to suggest other people to emulate here.

That said I am not so much against the current implementation and left my review as a comment intentionally.

**Comment by VictorVicente** [comment]
As Eric, I think we should keep the examples as simple to read as possible. Basically as a base guide on how to interact with the SDK for these features. 

I also like the idea of convert it to something like a CLI in the future, but I think we should redesign it a bit and think it a bit deeper, as many examples won't work on that way currently. 

We should also find if there is any real use case as a CLI first. 

In conclusion, really appreciate your initiative @sansan, I think it will trigger other discussions for these repo or as a completely new standalone tool but to keep consistency with other scripts, I'd prefer to keep them very simply showing how to use SDK for different purposes.
